h/services/bulk_annotation.py

```diff
@@ -0,0 +1,190 @@
+from typing import List, Union
+
+import sqlalchemy as sa
+from sqlalchemy.engine import Row
+from sqlalchemy.orm import Session
+from sqlalchemy.sql import Select
+
+from h.models import Annotation, AnnotationModeration, Group, User
+
+
+class BadDateFilter(ValueError):
+    """There is something wrong with the date filter provided."""
+
+
+class BadFieldSpec(ValueError):
+    """There is something wrong with the field you have specified."""
+
+
+def date_match(column, spec: dict):
+    """
+    Get an SQL comparator for a date column based on dict spec.
+
+    The dict can contain operators as keys and dates as values as per the
+    following complete (but nonsensical) filter:
+
+        {
+            "gt": "2012-11-30",
+            "gte": "2012-11-30",
+            "lt": "2012-11-30",
+            "lte": "2012-11-30",
+            "eq": "2012-11-30",
+            "ne": "2012-11-30",
+        }
+
+    :raises BadDateFilter: For unrecognised operators or no spec
```

Comment:
I think when we've discussed this before we've settled on `raises` because it reads more as an English sentence:

> "raises thing when this happens"

Our style is also loosely based on the examples from Sphinx:

 * https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists
 
This includes a list including all the styles we have in play, but the first choice and that given in the examples are `param` and `raises`.

Of course we can roll our own style if we want to, but it's not the way we've tended to go with other similar choices (like Black). There aren't many examples online using `arg` that I could find. `param` also happens to be the default in Pycharm which is nice for me as it automatically writes docstring templates for me.

In the case of `h` `raises` is the default way (23 instances vs 10). I think the situation is reversed in LMS.

Reply:
I'm happy to go with those (

> I think when we've discussed this before we've settled on `raises` because it reads more as an English sentence:
> 
> > "raises thing when this happens"

"raise thing when this happens" is surely also a sentence, in imperative tense.

I'm happy to go with your suggestions on these though. You could update https://stackoverflowteams.com/c/hypothesis/questions/92/134#134?